### PR TITLE
Cura 9277 cloud printer types

### PR DESCRIPTION
**Pull request by casperlamboo**

### CODE REVIEW

**Review by jellespijker — CHANGES_REQUESTED** [review]
Not yet tested but these are the comments from the code review. A couple of style and type hint issues.

But I'm mostly curious about the return value of the `createAbstractMachine` and why it isn't used

I will place this ticket back in to todo and performing some component testing in parallel

File: cura/Settings/AbstractMachine.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+class AbstractMachine(GlobalStack):
+    """ Represents a group of machines of the same type. This allows the user to select settings before selecting a printer. """
+
+    def __init__(self, container_id: str):
+        super().__init__(container_id)
+        self.setMetaDataEntry("type", "abstract_machine")
+
+    def getMachines(self) -> List[ContainerStack]:
+        from cura.CuraApplication import CuraApplication
```
[the remark this answers is not in the document]
Comment (Joeydelarago) [reply cura/Settings/AbstractMachine.py:17]: This is a bad copy paste from a @classmethod that needed this defined in the body

File: cura/Settings/AbstractMachine.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        application = CuraApplication.getInstance()
+        registry = application.getContainerRegistry()
+
+        printer_type = self.definition.getId()
+        cloud_printer_type = 3
```
[the remark this answers is not in the document]
Comment (Joeydelarago) [reply cura/Settings/AbstractMachine.py:22]: There is an intEnum in PrinterOutputDevice for this

File: cura/Settings/CuraStackBuilder.py [hunk 3]
```diff
@@ -199,17 +200,25 @@ def createGlobalStack(cls, new_stack_id: str, definition: DefinitionContainerInt
...
-        registry = application.getContainerRegistry()
 
-        stack = GlobalStack(new_stack_id)
-        stack.setDefinition(definition)
+        registry = application.getContainerRegistry()
 
         # Create user container
         user_container = cls.createUserChangesContainer(new_stack_id + "_user", definition.getId(), new_stack_id,
-                                                        is_global_stack = True)
+                                                        is_global_stack=True)
```
Comment (casperlamboo) [inline cura/Settings/CuraStackBuilder.py:221]:
```suggestion
                                                        is_global_stack = True)
```

File: cura/CuraApplication.py [hunk 4]
```diff
@@ -480,6 +484,7 @@ def __setLatestResouceVersionsForVersionUpgrade(self):
                 ("variant", InstanceContainer.Version * 1000000 + self.SettingVersion):                         (self.ResourceTypes.VariantInstanceContainer, "application/x-uranium-instancecontainer"),
                 ("setting_visibility", SettingVisibilityPresetsModel.Version * 1000000 + self.SettingVersion):  (self.ResourceTypes.SettingVisibilityPreset, "application/x-uranium-preferences"),
                 ("machine", 2):                                                                                 (Resources.DefinitionContainers, "application/x-uranium-definitioncontainer"),
+                ("abstract_machine", 2):                                                                        (Resources.DefinitionContainers, "application/x-uranium-definitioncontainer"),
```
[the remark this answers is not in the document]
Comment (casperlamboo) [reply cura/CuraApplication.py:487]:
```suggestion
                ("abstract_machine", 1):                                                                        (Resources.DefinitionContainers, "application/x-uranium-definitioncontainer"),
```

File: plugins/UM3NetworkPrinting/src/Cloud/CloudOutputDeviceManager.py [hunk 5]
```diff
@@ -405,6 +405,8 @@ def _createMachineFromDiscoveredDevice(self, key: str, activate: bool = True) ->
 
         self._setOutputDeviceMetadata(device, new_machine)
 
+        CuraStackBuilder.createAbstractMachine(device.printerType)
```
Comment (casperlamboo) [inline plugins/UM3NetworkPrinting/src/Cloud/CloudOutputDeviceManager.py:408]:
```suggestion
        _abstract_machine = CuraStackBuilder.createAbstractMachine(device.printerType)
```

File: cura/Settings/CuraStackBuilder.py [hunk 6]
```diff
@@ -259,3 +266,49 @@ def createDefinitionChangesContainer(cls, container_stack, container_name):
...
+        if registry.findContainerStacks(type="abstract_machine", id=abstract_machine_id):
+            # This abstract machine already exists
+            return None
+
+        match registry.findDefinitionContainers(type="machine", id=definition_id):
+            case []:
+                # It should not be possible for the definition to be missing since an abstract machine will only
+                # be created as a result of a machine with definition_id being created.
+                Logger.error("w", "Definition {definition} was not found!", definition=definition_id)
+                return None
```
[the remark this answers is not in the document]
Comment (casperlamboo) [reply cura/Settings/CuraStackBuilder.py:294]: https://github.com/Ultimaker/Cura/pull/13089/commits/12f3581337b434f26e120ab08e0bccedbe253777

File: cura/Settings/CuraStackBuilder.py [hunk 7]
```diff
@@ -259,3 +266,49 @@ def createDefinitionChangesContainer(cls, container_stack, container_name):
...
+                    application.empty_variant_container,
+                    application.empty_material_container,
+                    machine_node.preferredGlobalQuality().container,
+                )
+
+                stack.setName(name)
+
+                registry.addContainer(stack)
+
+                return stack
```
[the remark this answers is not in the document]
Comment (casperlamboo) [reply cura/Settings/CuraStackBuilder.py:314]: https://github.com/Ultimaker/Cura/pull/13089/commits/12f3581337b434f26e120ab08e0bccedbe253777

File: cura/Settings/CuraStackBuilder.py [hunk 8]
```diff
@@ -259,3 +266,49 @@ def createDefinitionChangesContainer(cls, container_stack, container_name):
...
+        abstract_machine_id = definition_id + "_abstract_machine"
+
+        from cura.CuraApplication import CuraApplication
+        application = CuraApplication.getInstance()
+        registry = application.getContainerRegistry()
+        container_tree = ContainerTree.getInstance()
+
+        if registry.findContainerStacks(type="abstract_machine", id=abstract_machine_id):
+            # This abstract machine already exists
+            return None
```
[the remark this answers is not in the document]
Comment (casperlamboo) [reply cura/Settings/CuraStackBuilder.py:287]: https://github.com/Ultimaker/Cura/pull/13089/commits/12f3581337b434f26e120ab08e0bccedbe253777

**Review by jellespijker — APPROVED** [review]

**Pull request merged**